File: src/retin/toolbox/document/Parameter.cpp
/*
Copyright © CNRS 2012. 
Authors: David Picard, Philippe-Henri Gosselin, Romain Negrel, Hedi 
Tabia
Contact: picard@ensea.fr

This software is governed by the CeCILL license under French law and
abiding by the rules of distribution of free software.  You can  use, 
modify and/ or redistribute the software under the terms of the CeCILL
license as circulated by CEA, CNRS and INRIA at the following URL
"http://www.cecill.info". 

As a counterpart to the access to the source code and rights to copy,
modify and redistribute granted by the license, users are provided only
with a limited warranty  and the software's author,  the holder of the
economic rights,  and the successive licensors  have only  limited
liability. 

In this respect, the user's attention is drawn to the risks associated
with loading,  using,  modifying and/or developing or reproducing the
software by the user in light of its specific status of free software,
that may mean  that it is complicated to manipulate,  and  that  also
therefore means  that it is reserved for developers  and  experienced
professionals having in-depth computer knowledge. Users are therefore
encouraged to load and test the software's suitability as regards their
requirements in conditions enabling the security of their systems and/or 
data to be ensured and,  more generally, to use and operate it in the 
same conditions as regards security. 

The fact that you are presently reading this means that you have had
knowledge of the CeCILL license and that you accept its terms.

*/
/**
 * \file Parameter.cpp
 * \author Philippe H. Gosselin
 * \version 4.0
 */

#include "retin/toolbox/document/Parameter.h"
#include "retin/toolbox/document/Parameters.h"

#include "retin/toolbox/document/serialization/Serializers.h" 

RETIN_SERIALIZABLE_OBJECT_DEFINITIONS(Parameter);

namespace retin  {

Parameter::Parameter (const Parameter& p) : value(p.value) {
	if (p.params)
	    params = boost::make_shared<Parameters>(*p.params);
}

bool	Parameter::equals(const Document* doc) const 
{
	if (!doc || !doc->isInstanceOf<Parameter>())
		return false; 
	const Parameter* x = static_cast<const Parameter*>(doc);
	if (x->value != value)
		return false;
	if (x->params) {
		if (!params)
			return false;
		return x->params->equals(params.get());
	}
	return !params;
}

}

File: src/retin/toolbox/document/Parameter.java
 /**
 * \file Parameter.java
 * \author Philippe H. Gosselin
 * \version 4.0
 */

package retin.toolbox.document;

import retin.toolbox.document.Parameters;

public class Parameter implements Comparable
{
	String		value;
	Parameters	params;

	public	Parameter ()
	{
	}

	public Parameter (String value)
	{
		this.value = value;
	}
	public Parameter (String value,Parameters params)
	{
		this.value = value;
		this.params = params;
	}

	public	String 	getValue ()
	{
		return value;
	}
	public	void	setValue (String value)
	{
		this.value = value;
	}

	public Parameters getParams ()
	{
		return params;
	}
	public	void	setParams (Parameters params)
	{
		this.params = params;
	}

	public	int	compareTo (Object o)
	{
		Parameter x = (Parameter)o;

	// Value
		if (value == null)
		{
			if (x.value != null)
				return -1;
		}
		else
		{
			if (x.value == null)
				return 1;
			int c = value.compareTo(x.value);
			if (c != 0) return c;
		}

	// Parameters
		if (params == null)
		{
			if (x.params != null)
				return -1;
		}
		else
		{
			if (x.params == null)
				return 1;
			int c = params.compareTo(x.params);
			if (c != 0) return c;
		}

		return 0;
	}
	public	boolean equals(Object o)
	{
		return compareTo(o) == 0;
	}

}
